### PR TITLE
Refactor of participants, ,play elimination

### PR DESCRIPTION
**Pull request by taahamahdi**
* Create Player upon VC entry instead of on-demand or elimination game start
* Remove redundant `participants` in GameSession and use scoreboard to manage instead
* Start elimination game with current VC members instead, and don't require to ,join
* Remove `Scoreboard::isEmpty()` and replace previous checks using it with checking to see if `Scoreboard::firstPlace` is empty (they were relying on the side effect that there are no scoreboard entries when there is no one in first place)
* Show players with 0 points on scoreboard, but only if they are in VC
* Provide more accurate counts of active players for stats by checking players are currently in VC
* Add `EliminationPlayer::fromUserID()`
* Remove redundant `Player::getID()`/`getName()` (already `public readonly`)

### CODE REVIEW

File: src/structures/game_session.ts [hunk 1]
```diff
@@ -1187,6 +1187,22 @@ export default class GameSession {
      * @param inVC - Whether the player is currently in the voice channel
      */
     setPlayerInVC(userID: string, inVC: boolean): void {
+        if (
+            inVC &&
+            !this.scoreboard.getPlayerIDs().includes(userID) &&
+            this.gameType !== GameType.TEAMS
+        ) {
+            this.scoreboard.addPlayer(
+                [GameType.CLASSIC, GameType.COMPETITION].includes(this.gameType)
```
Comment (Brainicism) [inline src/structures/game_session.ts:1196]: would it not making more sense to check `this.gameType == elimination`?

File: src/test/ci/begin.test.ts [hunk 2]
```diff
@@ -88,7 +53,7 @@ describe("begin command", () => {
                     assert.strictEqual(
                         BeginCommand.canStart(
                             gameSession,
-                            "1231",
+
```
[the remark this answers is not in the document]
Comment (taahamahdi) [reply src/test/ci/begin.test.ts:56]: Whoops, thanks

File: src/structures/game_session.ts (hunk 1 above)
Comment (taahamahdi) [reply src/structures/game_session.ts:1196]: You're right, thanks :]

File: src/structures/scoreboard.ts [hunk 3]
```diff
@@ -296,12 +289,18 @@ export default class Scoreboard {
...
      */
-    static getRanking(players: { [userID: string]: Player }): Array<string> {
-        return Object.entries(players)
-            .sort((a, b) => b[1].getScore() - a[1].getScore())
-            .map((x) => x[0]);
+    getRanking(): { [score: number]: number } {
+        const rankingToScore = {};
+        const sortedUniqueScores = [
+            ...new Set(Object.values(this.players).map((x) => x.getScore())),
+        ].sort((a, b) => b - a);
```
Comment (taahamahdi) [inline src/structures/scoreboard.ts:298]: Not sure if the sort is needed here anymore

File: src/test/ci/scoreboard.test.ts [hunk 4]
```diff
@@ -99,22 +101,102 @@ describe("score/exp updating", () => {
     describe("position changes", () => {
         it("should return the correct ranking of every player", async () => {
             const players = {
-                "12345": new Player("", "12345", "", 2),
+                ohmiID: new Player("", "ohmiID", "", 2),
```
Comment (Brainicism) [inline src/test/ci/scoreboard.test.ts:104]: Add tests for every player having different score, every player having same score

File: src/structures/player.ts [hunk 5]
```diff
@@ -42,24 +45,25 @@ export default class Player {
         this.avatarURL = avatarURL;
         this.expGain = 0;
         this.firstGameOfTheDay = firstGameOfTheDay;
+        this.previousRoundRanking = -1;
```
Comment (Brainicism) [inline src/structures/player.ts:48]: Why not `null`?

File: src/structures/scoreboard.ts (hunk 3 above)
Comment (Brainicism) [reply src/structures/scoreboard.ts:298]: Yes required

File: src/structures/scoreboard.ts [hunk 6]
```diff
@@ -296,12 +289,18 @@ export default class Scoreboard {
...
     /**
-     * @param players - user IDs mapped to Players
-     * @returns players IDs, sorted by rank in the game
+     * @returns a mapping of player scores to ranking
      */
-    static getRanking(players: { [userID: string]: Player }): Array<string> {
-        return Object.entries(players)
-            .sort((a, b) => b[1].getScore() - a[1].getScore())
-            .map((x) => x[0]);
+    getRanking(): { [score: number]: number } {
```
Comment (Brainicism) [inline src/structures/scoreboard.ts:294]: getScoreToRankingMap?

**Review by Brainicism — CHANGES_REQUESTED** [review]

**Review by Brainicism — APPROVED** [review]

**Pull request merged**